File: common/api/src/main/java/org/yechan/ApiResponseHttpMessageConverter.java
package org.yechan;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.time.Clock;
import java.time.LocalDateTime;
import org.jspecify.annotations.NullMarked;
import org.jspecify.annotations.Nullable;
import org.springframework.core.io.Resource;
import org.springframework.http.HttpInputMessage;
import org.springframework.http.HttpOutputMessage;
import org.springframework.http.MediaType;
import org.springframework.http.converter.AbstractHttpMessageConverter;
import org.springframework.http.converter.HttpMessageNotReadableException;
import org.springframework.http.converter.HttpMessageNotWritableException;
import org.springframework.http.server.ServletServerHttpResponse;
import org.springframework.web.servlet.mvc.method.annotation.StreamingResponseBody;
import tools.jackson.databind.ObjectMapper;

@NullMarked
class ApiResponseHttpMessageConverter extends AbstractHttpMessageConverter<Object> {

  private final ObjectMapper objectMapper;
  private final Clock clock;

  ApiResponseHttpMessageConverter(ObjectMapper objectMapper, Clock clock) {
    super(MediaType.APPLICATION_JSON, new MediaType("application", "*+json"));
    this.objectMapper = objectMapper;
    this.clock = clock;
  }

  @Override
  protected boolean supports(Class<?> clazz) {
    return !isExcludedType(clazz);
  }

  @Override
  public boolean canWrite(Class<?> clazz, @Nullable MediaType mediaType) {
    if (isExcludedType(clazz)) {
      return false;
    }

    return mediaType == null ||
        MediaType.ALL.equals(mediaType) || MediaType.APPLICATION_JSON.equals(mediaType)
        || mediaType.getSubtype().endsWith("+json");
  }

  @Override
  protected Object readInternal(Class<?> clazz, HttpInputMessage inputMessage)
      throws HttpMessageNotReadableException {
    throw new UnsupportedOperationException("응답 쓰기 전용 MessageConverter입니다.");
  }

  @Override
  protected void writeInternal(Object body, HttpOutputMessage outputMessage)
      throws IOException, HttpMessageNotWritableException {
    outputMessage.getHeaders().setContentType(MediaType.APPLICATION_JSON);

    int statusCode = resolveStatusCode(outputMessage);
    ApiResponse response = wrap(body, statusCode);

    objectMapper.writeValue(outputMessage.getBody(), response);
  }

  private ApiResponse wrap(Object body, int statusCode) {
    if (body instanceof ApiResponse response) {
      return response;
    }

    LocalDateTime now = LocalDateTime.now(clock);

    if (statusCode >= 400) {
      return new ApiErrorResponse(false, now, extractMessage(body));
    }
    return new ApiSuccessResponse<>(true, now, body);
  }

  private String extractMessage(Object body) {
    if (body instanceof String value) {
      return value;
    }
    if (body instanceof CharSequence value) {
      return value.toString();
    }
    if (body instanceof Throwable throwable) {
      if (throwable.getMessage() != null) {
        return throwable.getMessage();
      }
      return throwable.getClass().getSimpleName();
    }
    return String.valueOf(body);
  }

  private int resolveStatusCode(HttpOutputMessage outputMessage) {
    if (outputMessage instanceof ServletServerHttpResponse response) {
      return response.getServletResponse().getStatus();
    }
    return 200;
  }

  private boolean isExcludedType(Class<?> clazz) {
    return byte[].class.isAssignableFrom(clazz) ||
        ByteArrayOutputStream.class.isAssignableFrom(clazz) ||
        StreamingResponseBody.class.isAssignableFrom(clazz) ||
        Resource.class.isAssignableFrom(clazz);
  }
}

File: common/api/src/main/kotlin/org/yechan/ApiResponseHttpMessageConverter.kt
package org.yechan

import org.springframework.http.HttpInputMessage
import org.springframework.http.HttpOutputMessage
import org.springframework.http.MediaType
import org.springframework.http.converter.AbstractHttpMessageConverter
import org.springframework.http.server.ServletServerHttpResponse
import org.springframework.web.servlet.mvc.method.annotation.StreamingResponseBody
import tools.jackson.databind.ObjectMapper
import java.io.ByteArrayOutputStream
import java.time.Clock
import java.time.LocalDateTime

class ApiResponseHttpMessageConverter(
    private val objectMapper: ObjectMapper,
    private val clock: Clock,
) : AbstractHttpMessageConverter<Any>(
    MediaType.APPLICATION_JSON,
    MediaType("application", "*+json"),
) {
    override fun supports(clazz: Class<*>): Boolean = !isExcludedType(clazz)

    override fun canWrite(
        clazz: Class<*>,
        mediaType: MediaType?,
    ): Boolean {
        if (isExcludedType(clazz)) {
            return false
        }

        return mediaType == null ||
            mediaType == MediaType.ALL ||
            mediaType.isCompatibleWith(MediaType.APPLICATION_JSON) ||
            mediaType.subtype.endsWith("+json")
    }

    override fun readInternal(
        clazz: Class<out Any>,
        inputMessage: HttpInputMessage,
    ): Any = throw UnsupportedOperationException("응답 쓰기 전용 MessageConverter입니다.")

    override fun writeInternal(
        body: Any,
        outputMessage: HttpOutputMessage,
    ) {
        outputMessage.headers.contentType = MediaType.APPLICATION_JSON

        val statusCode = resolveStatusCode(outputMessage)
        val response = wrap(body, statusCode)

        objectMapper.writeValue(outputMessage.body, response)
    }

    private fun wrap(
        body: Any,
        statusCode: Int,
    ): ApiResponse {
        if (body is ApiResponse) {
            return body
        }

        val now = LocalDateTime.now(clock)

        return if (statusCode >= 400) {
            ApiErrorResponse(
                message = extractMessage(body),
                timestamp = now,
            )
        } else {
            ApiSuccessResponse(
                body = body,
                timestamp = now,
            )
        }
    }

    private fun extractMessage(body: Any): String = when (body) {
        is String -> body
        is CharSequence -> body.toString()
        is Throwable -> body.message ?: body::class.simpleName ?: "Unknown error"
        else -> body.toString()
    }

    private fun resolveStatusCode(outputMessage: HttpOutputMessage): Int = when (outputMessage) {
        is ServletServerHttpResponse -> outputMessage.servletResponse.status
        else -> 200
    }

    private fun isExcludedType(clazz: Class<*>): Boolean = ByteArray::class.java.isAssignableFrom(clazz) ||
        ByteArrayOutputStream::class.java.isAssignableFrom(clazz) ||
        StreamingResponseBody::class.java.isAssignableFrom(clazz) ||
        org.springframework.core.io.Resource::class.java.isAssignableFrom(clazz)
}
